File: backend/src/Nexora.Infrastructure/Services/NotificationService.cs
using Microsoft.AspNetCore.SignalR;
using Microsoft.Extensions.Logging;
using Nexora.Core.Interfaces;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace Nexora.Infrastructure.Services
{
    public class NotificationService : INotificationService
    {
        private readonly IHubContext<NotificationHub, INotificationHub> _hubContext;
        private readonly ILogger<NotificationService> _logger;
        private readonly Dictionary<string, UserConnection> _connections;
        private readonly List<NotificationHistory> _notificationHistory;

        public NotificationService(
            IHubContext<NotificationHub, INotificationHub> hubContext,
            ILogger<NotificationService> logger)
        {
            _hubContext = hubContext;
            _logger = logger;
            _connections = new Dictionary<string, UserConnection>();
            _notificationHistory = new List<NotificationHistory>();
        }

        public async Task SendPaymentStatusNotificationAsync(string tenantId, string userId, PaymentNotification notification)
        {
            try
            {
                _logger.LogInformation("Sending payment status notification for payment {PaymentId} to user {UserId}", 
                    notification.PaymentId, userId);

                var notificationData = new
                {
                    Type = "PaymentStatus",
                    Data = notification,
                    Timestamp = DateTime.UtcNow
                };

                await _hubContext.Clients.Group($"user_{userId}")
                    .SendAsync("PaymentStatusUpdate", notificationData);

                await _hubContext.Clients.Group($"tenant_{tenantId}")
                    .SendAsync("PaymentStatusUpdate", notificationData);

                await StoreNotificationHistory(tenantId, userId, NotificationType.PaymentStatus, 
                    "Payment Status Update", $"Payment {notification.PaymentId} status: {notification.Status}");

                _logger.LogInformation("Payment status notification sent successfully");
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Failed to send payment status notification");
                throw;
            }
        }

        public async Task SendSmsDeliveryNotificationAsync(string tenantId, string userId, SmsDeliveryNotification notification)
        {
            try
            {
                _logger.LogInformation("Sending SMS delivery notification for message {MessageId} to user {UserId}", 
                    notification.MessageId, userId);

                var notificationData = new
                {
                    Type = "SmsDelivery",
                    Data = notification,
                    Timestamp = DateTime.UtcNow
                };

                await _hubContext.Clients.Group($"user_{userId}")
                    .SendAsync("SmsDeliveryUpdate", notificationData);

                await _hubContext.Clients.Group($"tenant_{tenantId}")
                    .SendAsync("SmsDeliveryUpdate", notificationData);

                await StoreNotificationHistory(tenantId, userId, NotificationType.SmsDelivery, 
                    "SMS Delivery Update", $"SMS {notification.MessageId} status: {notification.Status}");

                _logger.LogInformation("SMS delivery notification sent successfully");
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Failed to send SMS delivery notification");
                throw;
            }
        }

        public async Task SendPaymentFailureAlertAsync(string tenantId, string userId, PaymentFailureAlert alert)
        {
            try
            {
                _logger.LogWarning("Sending payment failure alert for payment {PaymentId} to user {UserId}", 
                    alert.PaymentId, userId);

                var notificationData = new
                {
                    Type = "PaymentFailure",
                    Data = alert,
                    Timestamp = DateTime.UtcNow,
                    Priority = "High"
                };

                await _hubContext.Clients.Group($"user_{userId}")
                    .SendAsync("PaymentFailureAlert", notificationData);

                await _hubContext.Clients.Group($"tenant_{tenantId}_admins")
                    .SendAsync("PaymentFailureAlert", notificationData);

                await StoreNotificationHistory(tenantId, userId, NotificationType.PaymentFailure, 
                    "Payment Failure Alert", $"Payment {alert.PaymentId} failed: {alert.ErrorMessage}");

                _logger.LogInformation("Payment failure alert sent successfully");
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Failed to send payment failure alert");
                throw;
            }
        }

        public async Task SendBillingThresholdNotificationAsync(string tenantId, string userId, BillingThresholdNotification notification)
        {
            try
            {
                _logger.LogInformation("Sending billing threshold notification for {Service} to user {UserId}", 
                    notification.Service, userId);

                var notificationData = new
                {
                    Type = "BillingThreshold",
                    Data = notification,
                    Timestamp = DateTime.UtcNow,
                    Priority = notification.Severity.ToString()
                };

                await _hubContext.Clients.Group($"user_{userId}")
                    .SendAsync("BillingThresholdAlert", notificationData);

                await _hubContext.Clients.Group($"tenant_{tenantId}_billing")
                    .SendAsync("BillingThresholdAlert", notificationData);

                await StoreNotificationHistory(tenantId, userId, NotificationType.BillingThreshold, 
                    "Billing Threshold Alert", notification.Message);

                _logger.LogInformation("Billing threshold notification sent successfully");
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Failed to send billing threshold notification");
                throw;
            }
        }

        public async Task SendMultiChannelNotificationAsync(string tenantId, string userId, MultiChannelNotification notification)
        {
            try
            {
                _logger.LogInformation("Sending multi-channel notification {NotificationId} to user {UserId}", 
                    notification.NotificationId, userId);

                var notificationData = new
                {
                    Type = "MultiChannel",
                    Data = notification,
                    Timestamp = DateTime.UtcNow,
                    Priority = notification.Priority.ToString()
                };

                if (notification.Channels.Contains(NotificationChannel.WebSocket))
                {
                    await _hubContext.Clients.Group($"user_{userId}")
                        .SendAsync("MultiChannelNotification", notificationData);
                }

                await _hubContext.Clients.Group($"tenant_{tenantId}")
                    .SendAsync("MultiChannelNotification", notificationData);

                await StoreNotificationHistory(tenantId, userId, notification.Type, 
                    notification.Title, notification.Message);


                _logger.LogInformation("Multi-channel notification sent successfully");
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Failed to send multi-channel notification");
                throw;
            }
        }

        public async Task SubscribeToNotificationsAsync(string connectionId, string tenantId, string userId)
        {
            try
            {
                _logger.LogInformation("Subscribing connection {ConnectionId} for user {UserId} in tenant {TenantId}", 
                    connectionId, userId, tenantId);

                _connections[connectionId] = new UserConnection
                {
                    ConnectionId = connectionId,
                    TenantId = tenantId,
                    UserId = userId,
                    ConnectedAt = DateTime.UtcNow
                };

                await _hubContext.Groups.AddToGroupAsync(connectionId, $"user_{userId}");
                
                await _hubContext.Groups.AddToGroupAsync(connectionId, $"tenant_{tenantId}");

                _logger.LogInformation("Connection subscribed successfully");
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Failed to subscribe to notifications");
                throw;
            }
        }

        public async Task UnsubscribeFromNotificationsAsync(string connectionId)
        {
            try
            {
                if (_connections.TryGetValue(connectionId, out var connection))
                {
                    _logger.LogInformation("Unsubscribing connection {ConnectionId} for user {UserId}", 
                        connectionId, connection.UserId);

                    await _hubContext.Groups.RemoveFromGroupAsync(connectionId, $"user_{connection.UserId}");
                    await _hubContext.Groups.RemoveFromGroupAsync(connectionId, $"tenant_{connection.TenantId}");

                    _connections.Remove(connectionId);

                    _logger.LogInformation("Connection unsubscribed successfully");
                }
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Failed to unsubscribe from notifications");
                throw;
            }
        }

        public async Task<List<NotificationHistory>> GetNotificationHistoryAsync(string tenantId, string userId, int pageSize = 50, int pageNumber = 1)
        {
            try
            {
                _logger.LogInformation("Getting notification history for user {UserId} in tenant {TenantId}", 
                    userId, tenantId);

                var history = _notificationHistory
                    .Where(n => n.TenantId == tenantId && n.UserId == userId)
                    .OrderByDescending(n => n.CreatedAt)
                    .Skip((pageNumber - 1) * pageSize)
                    .Take(pageSize)
                    .ToList();

                return await Task.FromResult(history);
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Failed to get notification history");
                throw;
            }
        }

        private async Task StoreNotificationHistory(string tenantId, string userId, NotificationType type, string title, string message)
        {
            var notification = new NotificationHistory
            {
                Id = Guid.NewGuid().ToString(),
                TenantId = tenantId,
                UserId = userId,
                Type = type,
                Title = title,
                Message = message,
                Status = NotificationStatus.Sent,
                CreatedAt = DateTime.UtcNow
            };

            _notificationHistory.Add(notification);

            var userNotifications = _notificationHistory
                .Where(n => n.TenantId == tenantId && n.UserId == userId)
                .OrderByDescending(n => n.CreatedAt)
                .ToList();

            if (userNotifications.Count > 1000)
            {
                var toRemove = userNotifications.Skip(1000).ToList();
                foreach (var item in toRemove)
                {
                    _notificationHistory.Remove(item);
                }
            }

            await Task.CompletedTask;
        }

        private class UserConnection
        {
            public string ConnectionId { get; set; } = string.Empty;
            public string TenantId { get; set; } = string.Empty;
            public string UserId { get; set; } = string.Empty;
            public DateTime ConnectedAt { get; set; }
        }
    }
}

File: frontend/src/services/NotificationService.ts
import { HubConnection, HubConnectionBuilder, LogLevel } from '@microsoft/signalr';

export interface NotificationData {
  type: string;
  data: any;
  timestamp: string;
  priority?: string;
}

export interface PaymentNotification {
  paymentId: string;
  transactionId: string;
  status: 'Pending' | 'Processing' | 'Completed' | 'Failed' | 'Cancelled' | 'Refunded';
  amount: number;
  currency: string;
  paymentMethod: string;
  timestamp: string;
  message: string;
  metadata: Record<string, any>;
}

export interface SmsDeliveryNotification {
  messageId: string;
  phoneNumber: string;
  status: 'Queued' | 'Sent' | 'Delivered' | 'Failed' | 'Expired' | 'Rejected';
  cost: number;
  currency: string;
  provider: string;
  sentAt: string;
  deliveredAt?: string;
  errorMessage?: string;
  retryCount: number;
  metadata: Record<string, any>;
}

export interface PaymentFailureAlert {
  paymentId: string;
  transactionId: string;
  reason: 'InsufficientFunds' | 'InvalidCard' | 'ExpiredCard' | 'CardDeclined' | 'NetworkError' | 'ProcessingError' | 'FraudDetected' | 'LimitExceeded' | 'Unknown';
  errorCode: string;
  errorMessage: string;
  amount: number;
  currency: string;
  paymentMethod: string;
  timestamp: string;
  requiresAction: boolean;
  actionUrl?: string;
  metadata: Record<string, any>;
}

export interface BillingThresholdNotification {
  thresholdId: string;
  type: 'Usage' | 'Cost' | 'MessageCount' | 'ApiCalls';
  currentAmount: number;
  thresholdAmount: number;
  percentageUsed: number;
  currency: string;
  service: string;
  periodStart: string;
  periodEnd: string;
  severity: 'Info' | 'Warning' | 'Critical' | 'Emergency';
  message: string;
  actionUrl?: string;
  metadata: Record<string, any>;
}

export interface NotificationHistory {
  id: string;
  tenantId: string;
  userId: string;
  type: string;
  title: string;
  message: string;
  status: 'Pending' | 'Sent' | 'Delivered' | 'Read' | 'Failed' | 'Expired';
  createdAt: string;
  readAt?: string;
  metadata: Record<string, any>;
}

export type NotificationCallback = (notification: NotificationData) => void;

class NotificationService {
  private connection: HubConnection | null = null;
  private callbacks: Map<string, NotificationCallback[]> = new Map();
  private isConnected = false;
  private reconnectAttempts = 0;
  private maxReconnectAttempts = 5;
  private reconnectDelay = 1000;

  constructor() {
    this.setupConnection();
  }

  private setupConnection() {
    const token = localStorage.getItem('authToken');
    const apiUrl = process.env.REACT_APP_API_URL || 'https://localhost:7001';
    
    this.connection = new HubConnectionBuilder()
      .withUrl(`${apiUrl}/notificationHub`, {
        accessTokenFactory: () => token || '',
        withCredentials: true
      })
      .withAutomaticReconnect({
        nextRetryDelayInMilliseconds: (retryContext) => {
          if (retryContext.previousRetryCount < this.maxReconnectAttempts) {
            return Math.min(1000 * Math.pow(2, retryContext.previousRetryCount), 30000);
          }
          return null;
        }
      })
      .configureLogging(LogLevel.Information)
      .build();

    this.setupEventHandlers();
  }

  private setupEventHandlers() {
    if (!this.connection) return;

    this.connection.onclose((error) => {
      console.log('SignalR connection closed:', error);
      this.isConnected = false;
      this.notifyCallbacks('connectionClosed', { error });
    });

    this.connection.onreconnecting((error) => {
      console.log('SignalR reconnecting:', error);
      this.isConnected = false;
      this.notifyCallbacks('reconnecting', { error });
    });

    this.connection.onreconnected((connectionId) => {
      console.log('SignalR reconnected:', connectionId);
      this.isConnected = true;
      this.reconnectAttempts = 0;
      this.notifyCallbacks('reconnected', { connectionId });
    });

    this.connection.on('Connected', (data) => {
      console.log('Connected to notification hub:', data);
      this.isConnected = true;
      this.notifyCallbacks('connected', data);
    });

    this.connection.on('PaymentStatusUpdate', (notification: NotificationData) => {
      console.log('Payment status update:', notification);
      this.notifyCallbacks('paymentStatus', notification);
    });

    this.connection.on('SmsDeliveryUpdate', (notification: NotificationData) => {
      console.log('SMS delivery update:', notification);
      this.notifyCallbacks('smsDelivery', notification);
    });

    this.connection.on('PaymentFailureAlert', (notification: NotificationData) => {
      console.log('Payment failure alert:', notification);
      this.notifyCallbacks('paymentFailure', notification);
    });

    this.connection.on('BillingThresholdAlert', (notification: NotificationData) => {
      console.log('Billing threshold alert:', notification);
      this.notifyCallbacks('billingThreshold', notification);
    });

    this.connection.on('MultiChannelNotification', (notification: NotificationData) => {
      console.log('Multi-channel notification:', notification);
      this.notifyCallbacks('multiChannel', notification);
    });

    this.connection.on('NotificationHistory', (history: NotificationHistory[]) => {
      console.log('Notification history:', history);
      this.notifyCallbacks('notificationHistory', { history });
    });

    this.connection.on('NotificationMarkedAsRead', (data: { notificationId: string }) => {
      console.log('Notification marked as read:', data);
      this.notifyCallbacks('notificationRead', data);
    });
  }

  async connect(): Promise<void> {
    if (!this.connection || this.isConnected) return;

    try {
      await this.connection.start();
      console.log('SignalR connection established');
      this.isConnected = true;
      this.reconnectAttempts = 0;
    } catch (error) {
      console.error('Failed to connect to SignalR hub:', error);
      this.reconnectAttempts++;
      
      if (this.reconnectAttempts < this.maxReconnectAttempts) {
        setTimeout(() => this.connect(), this.reconnectDelay * this.reconnectAttempts);
      }
      
      throw error;
    }
  }

  async disconnect(): Promise<void> {
    if (!this.connection) return;

    try {
      await this.connection.stop();
      console.log('SignalR connection stopped');
      this.isConnected = false;
    } catch (error) {
      console.error('Error stopping SignalR connection:', error);
      throw error;
    }
  }

  subscribe(eventType: string, callback: NotificationCallback): void {
    if (!this.callbacks.has(eventType)) {
      this.callbacks.set(eventType, []);
    }
    this.callbacks.get(eventType)!.push(callback);
  }

  unsubscribe(eventType: string, callback: NotificationCallback): void {
    const callbacks = this.callbacks.get(eventType);
    if (callbacks) {
      const index = callbacks.indexOf(callback);
      if (index > -1) {
        callbacks.splice(index, 1);
      }
    }
  }

  private notifyCallbacks(eventType: string, data: any): void {
    const callbacks = this.callbacks.get(eventType);
    if (callbacks) {
      callbacks.forEach(callback => {
        try {
          callback({ type: eventType, data, timestamp: new Date().toISOString() });
        } catch (error) {
          console.error('Error in notification callback:', error);
        }
      });
    }
  }

  async markNotificationAsRead(notificationId: string): Promise<void> {
    if (!this.connection || !this.isConnected) {
      throw new Error('Not connected to notification hub');
    }

    try {
      await this.connection.invoke('MarkNotificationAsRead', notificationId);
    } catch (error) {
      console.error('Failed to mark notification as read:', error);
      throw error;
    }
  }

  async getNotificationHistory(pageSize: number = 50, pageNumber: number = 1): Promise<void> {
    if (!this.connection || !this.isConnected) {
      throw new Error('Not connected to notification hub');
    }

    try {
      await this.connection.invoke('GetNotificationHistory', pageSize, pageNumber);
    } catch (error) {
      console.error('Failed to get notification history:', error);
      throw error;
    }
  }

  async joinTenantGroup(tenantId: string): Promise<void> {
    if (!this.connection || !this.isConnected) {
      throw new Error('Not connected to notification hub');
    }

    try {
      await this.connection.invoke('JoinTenantGroup', tenantId);
    } catch (error) {
      console.error('Failed to join tenant group:', error);
      throw error;
    }
  }

  async leaveTenantGroup(tenantId: string): Promise<void> {
    if (!this.connection || !this.isConnected) {
      throw new Error('Not connected to notification hub');
    }

    try {
      await this.connection.invoke('LeaveTenantGroup', tenantId);
    } catch (error) {
      console.error('Failed to leave tenant group:', error);
      throw error;
    }
  }

  getConnectionState(): string {
    if (!this.connection) return 'Disconnected';
    return this.connection.state;
  }

  isConnectionActive(): boolean {
    return this.isConnected && this.connection?.state === 'Connected';
  }
}

export const notificationService = new NotificationService();
export default NotificationService;
